doc: stdcm: add documentation (#72)
DGEXSolutions/osrd-website: content/docs/reference/design-docs/stdcm/pathfinding_module/implementation_details.en.md

---
title: "Implementation details"
linkTitle: "6 - Implementation details"
weight: 60
---

This page is about implementation details.
It isn't necessary to understand general principles,
but it helps before reading the code.

#### STDCMEdgeBuilder

This class is used to make it easier to create instances of
`STDCMEdge`, the graph edges. Those contain many attributes,
most of which can be determined from the context (e.g. the
previous node).
The `STDCMEdgeBuilder` class makes some parameters optional
and automatically computes others.

Once instantiated and parameterd, an `STDCMEdgeBuilder` has two methods:

- `Collection<STDCMEdge> makeAllEdges()` can be used to create all
the possible edges in the given context for a given route.
If there are several "openings" between occupancy blocks, one edge
is instantiated for each opening. Every conflict, their avoidance,
and their related attributes are handled here.

- `STDCMEdge findEdgeSameNextOccupancy(double timeNextOccupancy)`:
This method is used to get the specific edges that uses a certain
opening (when it exists), identified here with the time of the next
occupancy block. It is called whenever a new edge must be re-created

> to replace an old one. It calls the previous method.

Reviewer: shouldn't this be in java code and just have a link to the file here ? What if classes are renamed at some point ? It will be hard to maintain in the future.

Author: It is in the java code, but I think this would still help before reading the code. It may be unclear where to start reading and where the logic is written.

It can be deleted if it becomes outdated and isn't worth updating.

Reviewer: Yes but since it is in different repository I'm almost sure that at some point it will be outdated and a reader will look for classes that won't be there anymore... If you don't want to do it this way then the java code should mention that if updated this page needs to be updated to. Maintaining two docs paraphrasing their content and keeping them up to date is a really difficult task

Author: I'm not too convinced this is necessary. If these methods are removed or significantly change in the future, it means the behavior has changed enough to rewrite most of the documentation. This isn't less stable than the rest of the documentation, they're really the "core" of the algorithm (which is why I want to talk about them).

### Pathfinding

#### Cost function

The function used to define pathfinding cost sets which path
is used over another. The result is always the one that minimizes
this cost (as long as the heuristic is admissible).

Here, two parameters are used: total run time and departure time.
The latter has a very small weight compared to the former,

> so that the fastest path is found. More details
> are explained in the documentation of those methods.

Reviewer: ```suggestion
so that the fastest path is found. More details
can be found in methods documentation <LINK>.
```

Author: Ideally we'd link the javadoc, but I don't know if/how we can do that. There's a javadoc file on the website but I don't know how to use it.

I'll link the .java file on github for now.
